src/array.cr

```diff
@@ -1909,7 +2021,67 @@ class Array(T)
   end
 
   private def check_needs_resize
-    double_capacity if @size == @capacity
+    # We have to compare against the actual capacity in case `@buffer` was moved
+    if @size == remaining_capacity
+      # If the array is not empty and more than half of the elements were shifted
+      # then we avoid a resize and just move the elements to the left.
+      # This is an heuristic. We could always try to move the elements if
+      # `@offset_to_buffer` is positive but it might happen that a user does
+      # `shift` + `push` in succession and it will produce a lot of memcopies.
+      if @capacity != 0 && @offset_to_buffer != 0 && @offset_to_buffer >= @capacity / 2
```

Comment:
`@offset_to_buffer != 0` is unnecessary.

Reply:
It is if capacity is 1. Maybe I should mention it.

Reply:
Clarified!